### DISCUSSION TITLE
Announcing function calling branch

### DISCUSSION THREAD

**Discussion by ahyatt** [ahyatt/llm#23] [Announcements]
Hi all,

After hearing that Ollama is going to add function calling, I was thinking it was time to add it this package.  I've now done that in the `function-calling` branch.  The README has the details.  

I'd love feedback on the design or utility.  Because not everything has function calling, I've implemented a new function `llm-capabilities` that enumerates what each LLM can do, so you can test before sending functions.

Everything function calling does is theoretically possible without it, but function calling should make getting structured information out of the text much easier.  And you can now let the LLM pilot Emacs, at least the functions that you convert from elisp into function-calling.  I have an example now that takes an elisp function and (via function calling) creates the function-called definition.  So now, in my example that works now, you can call a function to switch buffers based on an instruction.  This is very interesting to me, and I'm curious what would actually be useful to let the LLM do.

If there's no objections, I'll merge this into the llm main branch by next week, and make a new release.

**Comment by s-kostyaev** [comment]
Looks very interesting. I would like to try it once ollama will be supported.

**Reply by ahyatt** [reply]
It's possible that I could write an API call that will use function calling when available, and if not, basically get the same behavior in other ways.  That way we don't have to worry so much about compatibility.  Let me think about this.

**Reply by s-kostyaev** [reply]
Sure you can. See https://youtu.be/RXDWkiuXtG0?si=fzDYMigy8R9xYYMB
